Mathlib/Algebra/Order/Floor.lean

```diff
@@ -1720,61 +1720,61 @@ theorem subsingleton_floorRing {α} [LinearOrderedRing α] : Subsingleton (Floor
---     (ha : 0 < a) :
---     0 ≤ ⌊a⌋ :=
---   int_floor_nonneg ha.le
--- #align tactic.int_floor_nonneg_of_pos tactic.int_floor_nonneg_of_pos
-
--- /-- Extension for the `positivity` tactic: `Int.floor` is nonnegative if its input is. -/
--- @[positivity]
--- unsafe def positivity_floor : expr → tactic strictness
---   | q(⌊$(a)⌋) => do
---     let strictness_a ← core a
---     match strictness_a with
---       | positive p => nonnegative <$> mk_app `` int_floor_nonneg_of_pos [p]
---       | nonnegative p => nonnegative <$> mk_app `` int_floor_nonneg [p]
---       | _ => failed
---   | e => pp e >>= fail ∘ format.bracket "The expression `" "` is not of the form `⌊a⌋`"
--- #align tactic.positivity_floor tactic.positivity_floor
-
--- private theorem nat_ceil_pos [LinearOrderedSemiring α] [FloorSemiring α] {a : α} :
---     0 < a → 0 < ⌈a⌉₊ :=
---   Nat.ceil_pos.2
--- #align tactic.nat_ceil_pos tactic.nat_ceil_pos
-
--- private theorem int_ceil_pos [LinearOrderedRing α] [FloorRing α] {a : α} : 0 < a → 0 < ⌈a⌉ :=
---   Int.ceil_pos.2
--- #align tactic.int_ceil_pos tactic.int_ceil_pos
-
--- /-- Extension for the `positivity` tactic: `ceil` and `Int.ceil` are positive/nonnegative if
--- their input is. -/
--- @[positivity]
--- unsafe def positivity_ceil : expr → tactic strictness
---   | q(⌈$(a)⌉₊) => do
---     let positive p ← core a
---     -- We already know `0 ≤ n` for all `n : ℕ`
---         positive <$>
---         mk_app `` nat_ceil_pos [p]
---   | q(⌈$(a)⌉) => do
---     let strictness_a ← core a
---     match strictness_a with
---       | positive p => positive <$> mk_app `` int_ceil_pos [p]
---       | nonnegative p => nonnegative <$> mk_app `` Int.ceil_nonneg [p]
---       | _ => failed
---   | e => pp e >>=
---       fail ∘ format.bracket "The expression `" "` is not of the form `⌈a⌉₊` nor `⌈a⌉`"
--- #align tactic.positivity_ceil tactic.positivity_ceil
-
--- end Tactic
+namespace Mathlib.Meta.Positivity
+open Lean.Meta Qq
+
+private theorem int_floor_nonneg [LinearOrderedRing α] [FloorRing α] {a : α} (ha : 0 ≤ a) :
+    0 ≤ ⌊a⌋ :=
+  Int.floor_nonneg.2 ha
+
+private theorem int_floor_nonneg_of_pos [LinearOrderedRing α] [FloorRing α] {a : α}
+    (ha : 0 < a) :
+    0 ≤ ⌊a⌋ :=
+  int_floor_nonneg ha.le
+
+/-- Extension for the `positivity` tactic: `Int.floor` is nonnegative if its input is. -/
+@[positivity ⌊ _ ⌋]
+def evalFloor : PositivityExt where eval {u _α} _zα _pα e := do
+  -- match on `@Int.floor α' _ _ a`
+  let (.app (.app (.app (.app (.const ``Int.floor _) (α' : Q(Type $u))) _) _) a) ← whnfR e
+    | throwError "failed to match on Int.floor application"
+  let zα' : Q(Zero $α') ← synthInstanceQ q(Zero $α')
+  let pα' : Q(PartialOrder $α') ← synthInstanceQ q(PartialOrder $α')
+  match ← core zα' pα' a with
+  | .positive pa => pure (.nonnegative (← mkAppM ``int_floor_nonneg_of_pos #[pa]))
+  | .nonnegative pa => pure (.nonnegative (← mkAppM ``int_floor_nonneg #[pa]))
+  | _ => pure .none
```

Comment:
Here's a full Qq version:
```suggestion
/-- Extension for the `positivity` tactic: `Int.floor` is nonnegative if its input is. -/
@[positivity ⌊ _ ⌋]
def evalFloor : PositivityExt where eval {_u _α} _zα _pα e := do
  -- match on `@Int.floor α' _ _ a`
  have ez : Q(ℤ) := ← whnfR e
  let ~q(@Int.floor $α' $i $j $a) := ez | throwError "failed to match on Int.floor application"
  let zα' : Q(Zero $α') ← synthInstanceQ (u := u_1.succ) _
  let pα' : Q(PartialOrder $α') ← synthInstanceQ (u := u_1.succ) _
  assertInstancesCommute
  match ← core zα' pα' a with
  | .positive pa =>
      letI ret : Q(0 ≤ $ez) := q(int_floor_nonneg_of_pos (α := $α') $pa)
      pure (.nonnegative ret)
  | .nonnegative pa =>
      letI ret : Q(0 ≤ $ez) := q(int_floor_nonneg (α := $α') $pa)
      pure (.nonnegative ret)
  | _ => pure .none
```

Reply:
Thanks!
I actually prefer what I have, as it's shorter and easier (for me at least) to understand.

I guess that `u_1` gets inserted into the context from the `~q()` match? Is there a way to make that binding explicit?

Reply:
Yes, the `u_1` is being inserted unhygienically and I can't work out how to name it.

The advantage of my spelling is that it's much less fragile to argument changes to `Int.floor`.

Reply:
> The advantage of my spelling is that it's much less fragile to argument changes to `Int.floor`.

How so? It seems to me that your matching on `~q(@Int.floor $α' $i $j $a)` is essentially the same as my matching on `(.app (.app (.app (.app (.const ``Int.floor [u']) (α' : Q(Type $u'))) _) _) a)`

Reply:
Here's a shorter version
```suggestion
/-- Extension for the `positivity` tactic: `Int.floor` is nonnegative if its input is. -/
@[positivity ⌊ _ ⌋]
def evalFloor : PositivityExt where eval {_u _α} _zα _pα (e : Q(ℤ)) := do
  let ~q(@Int.floor $α' $i $j $a) := e | throwError "failed to match on Int.floor application"
  match ← core q(inferInstance) q(inferInstance) a with
  | .positive pa =>
      letI ret : Q(0 ≤ $e) := q(int_floor_nonneg_of_pos (α := $α') $pa)
      pure (.nonnegative <| ret)
  | .nonnegative pa =>
      letI ret : Q(0 ≤ $e) := q(int_floor_nonneg (α := $α') $pa)
      pure (.nonnegative ret)
  | _ => pure .none
```

Reply:
ah, that's starting to look more reasonable. I'll try that.

Reply:
> How so? It seems to me that your matching on `~q(@Int.floor $α' $i $j $a)` is essentially the same as my matching on `` (.app (.app (.app (.app (.const ``Int.floor [u']) (α' : Q(Type $u'))) _) _) a) ``

If the syntax changes, my code will become a Qq error at compile time. Yours will become a match failure at tactic use.